Run Cargo Tracker EE10 with Liberty blog (#3395)
OpenLiberty/blogs: posts/2023-10-20-open-liberty-with-cargo-tracker-ee10.adoc

---
layout: post

> title: "Running Eclipse Cargo Tracker, a Jakarta EE exemplar, with Liberty"

Reviewer: I see that the title and SEO title differ. I like the SEO title as it gives more context and key words like Open Liberty and Jakarta EE 10. Could we make these the same, or was there a reason this wasn't made the same?

Author: I applied the same rule for guides. The title should be clear. The seo title can contain more info, so that when searching from internet, the blog/guide can be found by more chances. It also applies to SEO description.

Reviewer: Ok sure, that makes sense. Thanks for the information. May be worth checking with Laura to get her thoughts on the title as I'm not an expert in SEO and search-ability.

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/gkwan-ibm
author_github: https://github.com/gkwan-ibm
seo-title: Run Eclipse Cargo Tracker, an exemplar Jakarta EE 10 application, with Open Liberty - OpenLiberty.io
seo-description: Experience Jakarta EE 10 and Open Liberty in an industrial environment by running Eclipse Cargo Tracker, an open-source application built to simulate real-world implementations of Jakarta EE 10.

> blog_description: "Experience Jakarta EE and Open Liberty in an industrial environment by running Eclipse Cargo Tracker, an open-source application built to simulate real-world implementations of Jakarta EE."

Reviewer: Maybe say "enterprise environment" instead of "industrial environment"?
Might also be nice to have Jakarta EE 10 at the start here of this sub-title for key word searches.

Author: what do you think if use "industrial standard"?

Reviewer: Still not a huge fan of the word industrial, but maybe worth checking with YK or Laura on this term?

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Running Eclipse Cargo Tracker, a Jakarta EE exemplar, with Liberty
Gilbert Kwan <https://github.com/gkwan-ibm>
:imagesdir: /
:url-prefix:
:url-about: /

> Recently, https://eclipse-ee4j.github.io/cargotracker/[Eclipse Cargo Tracker] was updated targeting Jakarta EE 10. The project also supports Open Liberty in addition to Payara and Eclipse GlassFish. Now, you can learn Jakarta EE in an industrial environment by running Cargo Tracker with Open Liberty, all on your own computer.

Reviewer: Not sure if "updated targeting Jakarta EE" is the right phrasing.
Maybe something like: "was updated to Jakarta EE 10" or "was updated to run on Jakarta EE 10"?

Reviewer: I would actually suggest changing this sentence to merge it with the second sentence. Perhaps something like:
"Just this month, the comprehensive Eclipse Cargo Tracker[LINK] exemplar application was updated with several key milestones, including targeting Jakarta EE 10 and support for key runtimes including Open Liberty as well as Payara and Eclipse Glassfish. This means that you can now learn all about Jakarta EE in an enterprise environment by running Cargo Tracker with Open Liberty, all on your local machine."

Author: How's about this?
> In September, the comprehensive [Eclipse Cargo Tracker](https://eclipse-ee4j.github.io/cargotracker/) exemplar application was updated with several key milestones, including targeting Jakarta EE 10 and support for key runtimes including Open Liberty as well as Payara and Eclipse Glassfish. This means that you can now learn all about Jakarta EE in an industrial standard by running Cargo Tracker with Open Liberty, all on your local machine.

Cargo Tracker simulates an industrially-developed freight tracking web application built with Jakarta EE 10 and domain-driven design principles. The application uses key Jakarta EE features, including Faces, CDI, Enterprise Beans, Bean Validation, RESTful Web Services, and JSON Binding. It also uses Persistence, along with Batch and Messaging.

Cargo Tracker books, tracks, updates, and routes cargo objects. Messaging events are logged to update cargo locations, while persistence ensures that cargo status is not changed or removed when the application is shut down. Each portion of Cargo Tracker represents a step in a real-life shipment process: the customer tracks their cargo using the Public Tracking interface, the shipping company schedules and routes deliveries using the Administration interface, and port handlers log arriving and departing cargo using the Mobile Event Logger. 

image::/img/blog/cargo-tracker-ee10-home.png[Cargo Tracker Home Page,width=50%,align="left"]
image::/img/blog/cargo-tracker-ee10-map.png[Cargo Tracker Map,width=50%,align="right"]

== Run Cargo Tracker with Open Liberty 

> Cargo Tracker EE10 supports Java SE 11 and 17. https://developer.ibm.com/languages/java/semeru-runtimes/downloads[IBM Semeru Runtime] is the preferred Java runtime when using Open Liberty. Before you clone the application to your machine, install any required JDKs and ensure that your `JAVA_HOME` system variable is set to a compatible version. To access the repository remotely you'll need to install https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[Git] if you haven't already. Finally, you will need to install https://maven.apache.org/download.cgi[Maven] to build and run the application. 

Reviewer: Is Cargo Tracker EE10 the official name of this new version. of the app? Is this correct?

Reviewer: Would perhaps reword the second sentence to be a little more introductory, as it feels like the audience should know what it is already at the moment.
Maybe something like: "When working with Open Liberty, you can make use of the IBM Semeru Runtime as your chosen Java runtime. This runtime offers performance benefits from deep technology investment in projects such as Eclipse OpenJ9 and is available across a wide variety of hardware and software platforms. To find out more about IBM Semeru Runtime, see our previous blog: Open Liberty and Semeru Runtimes, cloud-native performance that matters[https://openliberty.io/blog/2022/08/19/open-liberty-semeru-performance.html]."

Then I'd start a new line for the part of this paragraph that talks about "before you clone the app....".

You can clone the Cargo Tracker project to your local machine by running:

[source]
----
git clone https://github.com/eclipse-ee4j/cargotracker.git
----

You can also visit the https://github.com/eclipse-ee4j/cargotracker/tree/master[Git repository the Cargo Tracker project] to take a peek at the code.

To get started, choose one of the following setup methods:

* <<maven_setup, Maven command line setup>> 
* <<vscode_setup, Visual Studio Code IDE setup>>

> https://openliberty.io/docs/latest/development-mode.html[Liberty dev mode] is used in either setup because it provides three key features. Automatically detect, recompile, and deploy code changes in your running Liberty. Unit and integration tests are run on demand when you press `Enter` in the command terminal. Finally, you can attach a debugger to the running Liberty to step through the code. Through those features, you can deeply learn the Jarakee EE through debugging or updating the Cargo Tracker code.

Reviewer: "used in both setups" instead of either.
May be nice to make this a bulleted list instead of just prose - i.e. one for each feature.

Reviewer: remove the "the" before Jakarta EE

[#maven_setup]
== Maven command line setup

After cloning the project, you can quickly start Cargo Tracker through the command line with Maven. Navigate to the project base directory (where the pom.xml file is located) and run:

[source]
----
mvn -Popenliberty liberty:dev
----
 

> The application should start without any thrown exceptions. You can safely ignore the shrinkwrap features warning, message-driven bean warning, and the AggregateObjectMapping nested foreign key warning, as they don't affect the application functionality.

Reviewer: Can we link out to any external links that give further detail as to what these warnings are so if people are concerned they can look into them for themselves?

Author: I doubt we can

Reviewer: ok, no worries then.

image::/img/blog/cargo-tracker-ee10-dev-mode.png[Cargo Tracker Maven Output,width=70%,align="center"]

[#vscode_setup]
== Visual Studio Code IDE setup

> Ensure you are running Visual Studio Code (VSCode) IDE version 1.78.0 or later with Java 17. It is recommended to install Liberty Tools on your IDE from the https://marketplace.visualstudio.com/items?itemName=Open-Liberty.liberty-dev-vscode-ext[marketplace].

Reviewer: Can make this a little less formal. For the second sentence: "We'd also recommend installing the Liberty Tools for Visual Studio Code extension in your IDE to benefit from the fast and iterative development features it offers directly within the IDE. You can download this extension directly from the VS Code marketplace."

Reviewer: Should maybe also add that this extension is used in this demo/these screenshots/instructions as otherwise they won't have the dashboard used for the instructions after this.

Author: > Should maybe also add that this extension is used in this demo/these screenshots/instructions as otherwise they won't have the dashboard used for the instructions after this.

No sure what it means. Are you talking for the Liberty Tools doc to contain the screenshots?

After you start VSCode IDE, open the Cargo Tracker project by selecting **Explorer** > **Open Folder**. Select the `cargotracker` directory.

image::/img/blog/cargo-tracker-vscode-open-folder.png[Open Cargo Tracker project,width=70%,align="center"]

The `cargo-tracker` project appears under the `LIBERTY DASHBOARD` extension in the Explorer sidebar. To run Cargo Tracker, right-click the `cargo-tracker` project and select the **Start...** option.

image::/img/blog/cargo-tracker-vscode-start.png[Start Cargo Tracker,width=70%,align="center"]

You can then run dev mode with the Open Liberty profile active by typing in `-Popenliberty` in the pop-up entry field and then press the `enter/return` key. 

image::/img/blog/cargo-tracker-vscode-parameters.png[Start Cargo Tracker with dash P parameter,width=60%,align="center"]

> After starting with the custom parameter for the Open Liberty profile, a terminal window in the IDE opens and the Liberty Maven Plugin starts the application. Liberty is running in dev mode. You can safely ignore the shrinkwrap features warning, message-driven bean warning, and the AggregateObjectMapping nested foreign key warning, as they don't affect the application functionality.

Reviewer: Can we add a link to the Liberty Maven Plugin. Is this also a pre-req for this app to have as an extension installed in the IDE?

Reviewer: Also, same comment as above r.e. links for the warnings.

Author: Liberty Maven Plugin already defined in the pom.xml. It is a pre-req if run Cargo Tracker with Liberty

image::/img/blog/cargo-tracker-vscode-running.png[Cargo Tracker running in dev mode,width=70%,align="center"]

== Run unit tests

> To ensure the Cargo Tracker application started successfully, you can run the tests by pressing the `enter/return` key from the command-line session where you started dev mode. If the tests pass, you see a similar output to the following example:

Reviewer: "has started successfully" - missing the has

Reviewer: "you should see" - missing should. 
Could also be "will see"

image::/img/blog/cargo-tracker-ee10-unit-test-result.png[Cargo Tracker unit tests result,width=70%,align="center"]

== Access the application

> Once the application is running, you can access it through a browser of your choice at http://localhost:8080/cargo-tracker/ and start experimenting with it. You can track routed cargo through the Public Tracking interface, book and route cargo through the Administration interface, and update cargo object events using the Event Logger. You can stop the Liberty at any time by pressing `CTRL-C` from the command-line session where you started dev mode. 

Reviewer: "stop the Liberty server" - missing the word server.

Author: we do not mention Liberty as a server anymore. See https://github.com/OpenLiberty/guides-common/issues/944

Reviewer: Ok, that's fine. In which case, we need to remove the word "the" in this sentence as it doesn't make grammatical sense at the moment. So change it to "You can stop Liberty at any time"

image::/img/blog/cargo-tracker-ee10-dashboard-event-logger.png[Cargo Tracker Dashboard and Event Logger,width=60%,align="center"]

> For more information about of each of the interfaces, can check out the https://github.com/eclipse-ee4j/cargotracker/tree/master#exploring-the-application[Exploring the Application] section of the project README. 

Reviewer: "of the Cargo Tracker project README" - makes it explicitly clear which project readme we're referring to here.

== Nice work! Where to next?

> You've just set up Cargo Tracker and experienced Open Liberty with an enterprise-level application on your own machine! To learn the Cargo Tracker deeply, you can update the source code or use the debugger to step through the code.

Reviewer: "To understand the Cargo Tracker application in further detail, you can ...."

> Check out the https://openliberty.io/guides/[Open Liberty guides] for more information on using Jakarta EE and MicroProfile with Open Liberty. 

Reviewer: "for more information and interactive tutorials to set you through using more Jakarta EE and MicroProfile APIs with Open Liberty". Nice to include the fact that they are also interactive and more tutorial like rather than just documentation.

---
layout: post
title: "Running Eclipse Cargo Tracker, a Jakarta EE exemplar, with Liberty"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/gkwan-ibm
author_github: https://github.com/gkwan-ibm
seo-title: Run Eclipse Cargo Tracker, an exemplar Jakarta EE 10 application, with Open Liberty - OpenLiberty.io
seo-description: Experience Jakarta EE 10 and Open Liberty in an industrial environment by running Eclipse Cargo Tracker, an open-source application built to simulate real-world implementations of Jakarta EE 10.
blog_description: "Experience Jakarta EE and Open Liberty in an industrial environment by running Eclipse Cargo Tracker, an open-source application built to simulate real-world implementations of Jakarta EE."
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Running Eclipse Cargo Tracker, a Jakarta EE exemplar, with Liberty
Gilbert Kwan <https://github.com/gkwan-ibm>
:imagesdir: /
:url-prefix:
:url-about: /

> In September, the comprehensive https://eclipse-ee4j.github.io/cargotracker/[Eclipse Cargo Tracker] exemplar application was updated with several key milestones, including targeting Jakarta EE 10 and adding support for key runtimes including Open Liberty as well as Payara and Eclipse Glassfish. This means that you can now learn all about Jakarta EE by running Cargo Tracker with Open Liberty, all on your local machine.

Reviewer 2: ```suggestion
In September, the comprehensive https://eclipse-ee4j.github.io/cargotracker/[Eclipse Cargo Tracker] exemplar application was updated to target Jakarta EE 10 and to support for key runtimes, including Open Liberty. This means you can now learn about Jakarta EE by running Cargo Tracker with Open Liberty, all on your local machine.
```

Cargo Tracker simulates an industrially-developed freight tracking web application built with Jakarta EE 10 and domain-driven design principles. The application uses key Jakarta EE features, including Faces, CDI, Enterprise Beans, Bean Validation, RESTful Web Services, and JSON Binding. It also uses Persistence, along with Batch and Messaging.

Cargo Tracker books, tracks, updates, and routes cargo objects. Messaging events are logged to update cargo locations, while persistence ensures that cargo status is not changed or removed when the application is shut down. Each portion of Cargo Tracker represents a step in a real-life shipment process: the customer tracks their cargo using the Public Tracking interface, the shipping company schedules and routes deliveries using the Administration interface, and port handlers log arriving and departing cargo using the Mobile Event Logger. 

image::/img/blog/cargo-tracker-ee10-home.png[Cargo Tracker Home Page,width=50%,align="left"]
image::/img/blog/cargo-tracker-ee10-map.png[Cargo Tracker Map,width=50%,align="right"]

== Run Cargo Tracker with Open Liberty 

> Cargo Tracker supports Java SE 11 and 17. When working with Open Liberty, you can make use of the https://developer.ibm.com/languages/java/semeru-runtimes/downloads[IBM Semeru Runtime] as your chosen Java runtime. This runtime offers performance benefits from deep technology investment in projects such as Eclipse OpenJ9 and is available across a wide variety of hardware and software platforms. To find out more about IBM Semeru Runtime, see our previous blog: https://openliberty.io/blog/2022/08/19/open-liberty-semeru-performance.html[Open Liberty and Semeru Runtimes, cloud-native performance that matters].

Reviewer 2: ```suggestion
Cargo Tracker supports Java SE 11 and 17. When working with Open Liberty, you can make use of the https://developer.ibm.com/languages/java/semeru-runtimes/downloads[IBM Semeru Runtime] as your chosen Java runtime. This runtime offers performance benefits from deep technology investment in projects such as Eclipse OpenJ9 and is available across a wide variety of hardware and software platforms. To find out more about IBM Semeru Runtime, see https://openliberty.io/blog/2022/08/19/open-liberty-semeru-performance.html[Open Liberty and Semeru Runtimes, cloud-native performance that matters].
```

== Access the application

Once the application is running, you can access it through a browser of your choice at http://localhost:8080/cargo-tracker/ and start experimenting with it. You can track routed cargo through the Public Tracking interface, book and route cargo through the Administration interface, and update cargo object events using the Event Logger. You can stop Liberty at any time by pressing `CTRL-C` from the command-line session where you started dev mode. 

image::/img/blog/cargo-tracker-ee10-dashboard-event-logger.png[Cargo Tracker Dashboard and Event Logger,width=60%,align="center"]

For more information about of each of the interfaces, can check out the https://github.com/eclipse-ee4j/cargotracker/tree/master#exploring-the-application[Exploring the Application] section of the Cargo Tracker project README. 

== Nice work! Where to next?

You've just set up Cargo Tracker and experienced Open Liberty with an enterprise-level application on your own machine! To understand the Cargo Tracker application in further detail, you can update the source code or use the debugger to step through the code.

> Check out the https://openliberty.io/guides/[Open Liberty guides] for more information and interactive tutorials to set you through using more Jakarta EE and MicroProfile APIs with Open Liberty. 

Reviewer 2: ```suggestion
Check out the https://openliberty.io/guides/[Open Liberty guides] for more information and interactive tutorials that walk you through using more Jakarta EE and MicroProfile APIs with Open Liberty. 
```

---
layout: post
title: "Running Eclipse Cargo Tracker, a Jakarta EE exemplar, with Liberty"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/gkwan-ibm
author_github: https://github.com/gkwan-ibm
seo-title: Run Eclipse Cargo Tracker, an exemplar Jakarta EE 10 application, with Open Liberty - OpenLiberty.io
seo-description: Experience Jakarta EE 10 and Open Liberty in an industrial environment by running Eclipse Cargo Tracker, an open-source application built to simulate real-world implementations of Jakarta EE 10.
blog_description: "Experience Jakarta EE and Open Liberty in an industrial environment by running Eclipse Cargo Tracker, an open-source application built to simulate real-world implementations of Jakarta EE."
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Running Eclipse Cargo Tracker, a Jakarta EE exemplar, with Liberty
Gilbert Kwan <https://github.com/gkwan-ibm>
:imagesdir: /
:url-prefix:
:url-about: /

> In September, the comprehensive https://eclipse-ee4j.github.io/cargotracker/[Eclipse Cargo Tracker] exemplar application was updated with several key milestones, including targeting Jakarta EE 10 and support for key runtimes including Open Liberty as well as Payara and Eclipse Glassfish. This means that you can now learn all about Jakarta EE in an industrial standard by running Cargo Tracker with Open Liberty, all on your local machine.

Reviewer: "and adding support for key runtimes" - missing adjective here, I've added "adding"

Reviewer: Still not sure on "industrial standard", perhaps we can ask YK or Laura on this?

Author: will ask after your review is completed

Author: How's about remove the phase "in an industrial standard"?

> In September, the comprehensive https://eclipse-ee4j.github.io/cargotracker/[Eclipse Cargo Tracker] exemplar application was updated with several key milestones, including targeting Jakarta EE 10 and adding support for key runtimes including Open Liberty as well as Payara and Eclipse Glassfish. This means that you can now learn all about Jakarta EE by running Cargo Tracker with Open Liberty, all on your local machine.

== What is the Cargo Tracker application?

> As one of the top Java framework choices for enterprise-level and cloud-native Java applications, Jakarta EE is at the center of projects all over the globe. The Cargo Tracker open-source application was designed to replicate a complex enterprise application by using core Jakarta EE features with professional development standards.

Reviewer: I would remove level and and, so instead have "for cloud-native enterprise Java applications"

Reviewer: Switch ordering in next sentence to: "The open-source Cargo Tracker application"

Before you clone the application to your machine, install any required JDKs and ensure that your `JAVA_HOME` system variable is set to a compatible version. To access the repository remotely you'll need to install https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[Git] if you haven't already. Finally, you will need to install https://maven.apache.org/download.cgi[Maven] to build and run the application. 

You can clone the Cargo Tracker project to your local machine by running:

[source]
----
git clone https://github.com/eclipse-ee4j/cargotracker.git
----

You can also visit the https://github.com/eclipse-ee4j/cargotracker/tree/master[Git repository the Cargo Tracker project] to take a peek at the code.

To get started, choose one of the following setup methods:

* <<maven_setup, Maven command line setup>> 
* <<vscode_setup, Visual Studio Code IDE setup>>

> https://openliberty.io/docs/latest/development-mode.html[Liberty dev mode] is used in both setup. Because it provides the following three key features, you can deeply learn Jarakee EE through debugging or updating the Cargo Tracker code:

Reviewer: both setups - missing s here

Reviewer: Remove Because in next sentence, so just start with "Dev mode provides the following three key features, enabling you to quickly and easily develop your experience with Jakarta EE through the Cargo Tracker application:"

We'd also recommend installing the Liberty Tools for Visual Studio Code extension in your IDE to benefit from the fast and iterative development features it offers directly within the IDE. You can download this extension directly from the https://marketplace.visualstudio.com/items?itemName=Open-Liberty.liberty-dev-vscode-ext[VS Code marketplace].

After you start VSCode IDE, open the Cargo Tracker project by selecting **Explorer** > **Open Folder**. Select the `cargotracker` directory.

image::/img/blog/cargo-tracker-vscode-open-folder.png[Open Cargo Tracker project,width=70%,align="center"]

The `cargo-tracker` project appears under the `LIBERTY DASHBOARD` extension in the Explorer sidebar. To run Cargo Tracker, right-click the `cargo-tracker` project and select the **Start...** option.

image::/img/blog/cargo-tracker-vscode-start.png[Start Cargo Tracker,width=70%,align="center"]

You can then run dev mode with the Open Liberty profile active by typing in `-Popenliberty` in the pop-up entry field and then press the `enter/return` key. 

image::/img/blog/cargo-tracker-vscode-parameters.png[Start Cargo Tracker with dash P parameter,width=60%,align="center"]

> After starting with the custom parameter for the Open Liberty profile, a terminal window in the IDE opens and the https://github.com/OpenLiberty/ci.maven#readme[Liberty Maven plug-in] starts the application. Liberty is running in dev mode. You can safely ignore the shrinkwrap features warning, message-driven bean warning, and the AggregateObjectMapping nested foreign key warning, as they don't affect the application functionality.

Reviewer: Liberty is now running in dev mode - maybe add now to show how it links to previous sentence?
